libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,378 @@
+		return result;
+	}
+
+	using evmasm::SemanticInformation;
+
+	return util::applyMap(
+		SemanticInformation::readWriteOperations(*instruction),
+		[&](SemanticInformation::Operation const& _op) -> Operation
+		{
+			yulAssert(!(_op.lengthParameter && _op.lengthConstant));
+			yulAssert(_op.effect != Effect::None);
+			Operation ourOp{_op.location, _op.effect, {}, {}};
+			if (_op.startParameter)
+				ourOp.start = identifierNameIfSSA(_functionCall.arguments.at(*_op.startParameter));
+			if (_op.lengthParameter)
+				ourOp.length = identifierNameIfSSA(_functionCall.arguments.at(*_op.lengthParameter));
+			if (_op.lengthConstant)
+				switch (*_op.lengthConstant)
+				{
+				case 1: ourOp.length = YulString(one); break;
+				case 32: ourOp.length = YulString(thirtyTwo); break;
+				default: yulAssert(false);
+				}
+			return ourOp;
+		}
+	);
+}
+
+void UnusedStoreEliminator::applyOperation(UnusedStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool UnusedStoreEliminator::knownUnrelated(
+	UnusedStoreEliminator::Operation const& _op1,
+	UnusedStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+		{
+			yulAssert(
+				_op1.length &&
+				_op2.length &&
+				knowledge.valueIfKnownConstant(*_op1.length) == 1 &&
+				knowledge.valueIfKnownConstant(*_op2.length) == 1
+			);
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+		}
+	}
+	else
+	{
+		yulAssert(_op1.location == Location::Memory, "");
+		if (
+			(_op1.length && knowledge.knownToBeZero(*_op1.length)) ||
+			(_op2.length && knowledge.knownToBeZero(*_op2.length))
+		)
+			return true;
+
+		u256 largeNumber = u256(1) << 255;
+		// Due to wrapping arithmetic, the below code can have "false positives".
+		// In the situation that the subtraction for computing "diff" wraps,
+		// we can return "unrelated" although the writes are actually related.
+		// In this case, 2.start and 1.start will both be >= largeNumber
+		// (see test/formal/redundant_store_unrelated.py).
+		// This means, both operations revert.
+		// Because of these false positives, this optimizer routine removes write operations
+		// where all subsequent read operations are unrelated or will revert due to high memory
+		// access. Since this second class of read operations will never return a value,
+		// it is irrelevant what value they return, so the write can be safely removed.
```

Comment:
It was able to optimize this out: https://github.com/ethereum/solidity/pull/12672/files#diff-454ca1e7741b290c37ead677fb52b645bd9c18325aa1aa5d023df0b926b2e1b7R16

I'm not sure what the impact in practice was. Of course, "known unrelated" is always difficult because we need the code to also terminate in the same function in order to say something about memory.

Still, if we want to improve this with the linear solver, we need exactly the same reasoning here.

Reply:
I just checked, that is the only change for isoltest.

Reply:
I would have hoped that the linear solver could actually tell us stuff like if ``_op1.start <= _op2.start`` and then the problem just vanishes...

Reply:
But it might not be able to do that accurately either if either of them could ever have overflowed...

Reply:
It still seems a bit weird to me to have the optimizer operate under conflicting assumptions - we say both "large memory reads don't return values", but we also allow them to be load-resolved, so we also say "we can assume any memory read to return a value"...
I don't see any issue due to this in practice (yet?), but it will never be formally correct like this at least :-)...

Reply:
And I actually wouldn't be surprised there was a case, in which this actually happens:
- a store to high memory location A
- a store to high memory location B
- a read from high memory location A

in which B is spuriously unrelated to A... then the write to B would be removed and we may resolve the read from A, after which we can potentially further remove the write to A...
Even if that doesn't happen the way things are set up right now for any concrete values of A and B, we'd have the constant danger of this happening, once we extend the optimizer anywhere...

Reply:
Ok, the interaction with the load resolver is actually a good point.

Reply:
To do this stuff properly, maybe there is no way around to proving that the memory pointer and arithmetic on it never overflows... but that's probably always going to be tricky (especially across functions calls and such)...

Reply:
On the other hand: any code we generate in which the free memory pointer does overflow is ill-defined anyways (since we may then remove all kinds of memory writes and reads, but have a write after the memory pointer overflow mess with the scratch space or the free memory pointer itself)...